File: jdi framework/jdi-uitest-web/src/main/java/com/epam/jdi/uitests/web/selenium/elements/WebCascadeInit.java
package com.epam.jdi.uitests.web.selenium.elements;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */


import com.epam.jdi.commons.pairs.Pairs;
import com.epam.jdi.uitests.core.interfaces.CascadeInit;
import com.epam.jdi.uitests.core.interfaces.base.IBaseElement;
import com.epam.jdi.uitests.core.interfaces.base.IComposite;
import com.epam.jdi.uitests.core.interfaces.complex.IPage;
import com.epam.jdi.uitests.web.selenium.elements.apiInteract.ContextType;
import com.epam.jdi.uitests.web.selenium.elements.complex.Elements;
import com.epam.jdi.uitests.web.selenium.elements.composite.WebPage;
import com.epam.jdi.uitests.web.selenium.elements.pageobjects.annotations.Frame;
import com.epam.jdi.uitests.web.selenium.elements.pageobjects.annotations.JFindBy;
import com.epam.jdi.uitests.web.selenium.elements.pageobjects.annotations.JPage;
import org.openqa.selenium.By;
import org.openqa.selenium.support.FindBy;

import java.lang.reflect.Field;
import java.lang.reflect.ParameterizedType;
import java.util.List;

import static com.epam.jdi.commons.ReflectionUtils.*;
import static com.epam.jdi.commons.StringUtils.LINE_BREAK;
import static com.epam.jdi.uitests.core.settings.JDIData.appVersion;
import static com.epam.jdi.uitests.core.settings.JDISettings.exception;
import static com.epam.jdi.uitests.web.selenium.elements.pageobjects.annotations.WebAnnotationsUtil.*;
import static java.lang.String.format;

/**
 * Created by Roman_Iovlev on 6/10/2015.
 */
public class WebCascadeInit extends CascadeInit {

    protected void setElement(Class<?> parentType, Field field, String driverName) {
        try {
            Class<?> type = field.getType();
            IBaseElement instance;
            if (isInterface(type, IPage.class)) {
                instance = (IBaseElement) getValueField(field, null);
                if (instance == null)
                    instance = (IBaseElement) type.newInstance();
                fillPage(instance, field, parentType);
            } else {
                instance = createChildFromFieldStatic(parentType, field, type, driverName);
                instance.setFunction(getFunction(field));
            }
            instance.setName(field);
            if (instance.getClass().getSimpleName().equals(""))
                instance.setTypeName(type.getSimpleName());
            instance.setParentName(parentType.getClass().getSimpleName());
            field.set(null, instance);
            if (isInterface(field, IComposite.class))
                initElements(instance, driverName);
        } catch (Exception ex) {
            throw exception("Error in setElement for field '%s' with parent '%s'", field.getName(), parentType.getClass().getSimpleName() + LINE_BREAK + ex.getMessage());
        }
    }

    protected void setElement(Object parent, Field field, String driverName) {
        try {
            Class<?> type = field.getType();
            IBaseElement instance;
            if (isInterface(type, IPage.class)) {
                instance = (IBaseElement) getValueField(field, parent);
                if (instance == null)
                    instance = (IBaseElement) type.newInstance();
                fillPage(instance, field, parent != null ? parent.getClass() : null);
            } else {
                instance = createChildFromField(parent, field, type, driverName);
                instance.setFunction(getFunction(field));
            }
            instance.setName(field);
            instance.getAvatar().setDriverName(driverName);
            if (instance.getClass().getSimpleName().equals(""))
                instance.setTypeName(type.getSimpleName());
            instance.setParentName(getClassName(parent));
            field.set(parent, instance);
            if (isInterface(field, IComposite.class))
                initElements(instance, driverName);
        } catch (Exception ex) {
            throw exception("Error in setElement for field '%s' with parent '%s'", field.getName(),
                    getClassName(parent) + LINE_BREAK + ex.getMessage());
        }
    }

    private void fillPage(IBaseElement instance, Field field, Class<?> parentType) {
        if (field.isAnnotationPresent(JPage.class))
            fillPageFromAnnotaiton((WebPage) instance, field.getAnnotation(JPage.class), parentType);
    }

    private BaseElement createChildFromFieldStatic(Class<?> parentClass, Field field, Class<?> type, String driverName) {
        BaseElement instance = (BaseElement) getValueField(field, null);
        if (instance == null)
            try {
                instance = getElementInstance(field, driverName);
            } catch (Exception ex) {
                throw exception(format("Can't create child for parent '%s' with type '%s'",
                        parentClass.getSimpleName(), field.getType().getSimpleName()));
            }
        else if (instance.getLocator() == null)
            instance.avatar.byLocator = getNewLocator(field);
        instance.avatar.context = new Pairs<>();
        By frameBy = getFrame(field.getDeclaredAnnotation(Frame.class));
        if (frameBy != null)
            instance.avatar.context.add(ContextType.Frame, frameBy);
        return instance;
    }

    private IBaseElement createChildFromField(Object parent, Field field, Class<?> type, String driverName) {
        BaseElement instance = (BaseElement) getValueField(field, parent);
        if (instance == null)
            try {
                instance = getElementInstance(field, driverName);
            } catch (Exception ex) {
                throw exception(
                        format("Can't create child for parent '%s' with type '%s'",
                                parent.getClass().getSimpleName(), field.getType().getSimpleName()));
            }
        else if (instance.getLocator() == null)
            instance.avatar.byLocator = getNewLocator(field);
        instance.avatar.context = (isBaseElement(parent))
                ? ((BaseElement) parent).avatar.context.copy()
                : new Pairs<>();
        if (type != null) {
            By frameBy = getFrame(field.getDeclaredAnnotation(Frame.class));
            if (frameBy != null)
                instance.avatar.context.add(ContextType.Frame, frameBy);
        }
        if (isBaseElement(parent)) {
            By parentLocator = ((BaseElement) parent).getLocator();
            if (parentLocator != null)
                instance.avatar.context.add(ContextType.Locator, parentLocator);
        }
        return instance;
    }

    private BaseElement getElementInstance(Field field, String driverName) {
        Class<?> type = field.getType();
        String fieldName = field.getName();
        By newLocator = getNewLocator(field);
        try {
            BaseElement instance = null;
            if (isClass(type, List.class)) {
                Class<?> elementClass = (Class<?>) ((ParameterizedType) field.getGenericType()).getActualTypeArguments()[0];
                if (elementClass.isInterface())
                    elementClass = MapInterfaceToElement.getClassFromInterface(type);
                if (elementClass != null)
                    instance = new Elements(newLocator, elementClass);
            } else {
                if (type.isInterface())
                    type = MapInterfaceToElement.getClassFromInterface(type);
                if (type != null) {
                    instance = (BaseElement) type.newInstance();
                    instance.avatar.byLocator = newLocator;
                }
            }
            if (instance != null) {
                instance.avatar.setDriverName(driverName);
                return instance;
            }
            throw exception("Unknown interface: " + type
                    + ". Add relation interface -> class in VIElement.InterfaceTypeMap");
        } catch (Exception ex) {
            throw exception("Error in getElementInstance for field '%s' with type '%s'", fieldName, type.getSimpleName()
                    + LINE_BREAK + ex.getMessage());
        }
    }

    private By getNewLocator(Field field) {
        try {
            By byLocator = null;
            String locatorGroup = appVersion;
            if (locatorGroup != null) {
                JFindBy jFindBy = field.getAnnotation(JFindBy.class);
                if (jFindBy != null && locatorGroup.equals(jFindBy.group()))
                    byLocator = getFindByLocator(jFindBy);
            }
            return (byLocator != null)
                    ? byLocator
                    : getFindByLocator(field.getAnnotation(FindBy.class));
        } catch (Exception ex) {
            throw exception("Error in get locator for type '%s'", field.getType().getName()
                    + LINE_BREAK + ex.getMessage());
        }
    }

}

File: C#.Net/JDI/Web/Selenium/Base/WebCascadeInit.cs
using System;
using System.Collections;
using System.Reflection;
using Epam.JDI.Commons;
using Epam.JDI.Core.Base;
using Epam.JDI.Core.Interfaces.Base;
using Epam.JDI.Core.Interfaces.Complex;
using Epam.JDI.Web.Attributes;
using Epam.JDI.Web.Selenium.Elements.Complex;
using Epam.JDI.Web.Selenium.Elements.Composite;
using Epam.JDI.Web.Settings;
using NUnit.Framework;
using OpenQA.Selenium;
using RestSharp.Extensions;
using static Epam.JDI.Core.ExceptionUtils;
using static Epam.JDI.Core.Attributes.AttributesUtil;
using static Epam.JDI.Core.Settings.JDIData;
using static Epam.JDI.Core.Settings.JDISettings;

namespace Epam.JDI.Web.Selenium.Base
{
    public class WebCascadeInit : CascadeInit
    {
        protected override void SetElement(object parent, Type parentType, FieldInfo field, string driverName)
        {
            ActionWithException(() =>
            {
                var type = field.FieldType;
                var instance = typeof(IPage).IsAssignableFrom(type)
                    ? GetInstancePage(parent, field, type, parentType)
                    : GetInstanceElement(parent, parentType, field, driverName);
                instance.SetName(field);
                instance.Avatar.DriverName = driverName;
                instance.TypeName = type.Name;
                instance.Parent = parent;
                field.SetValue(parent, instance);
                if (typeof(IComposite).IsAssignableFrom(type))
                    InitElements(instance, driverName);
            }, ex => $"Error in setElement for field '{field.Name}' with parent '{parentType?.Name ?? "NULL Class" + ex.FromNewLine()}'");
        }

        private static IBaseElement GetInstancePage(object parent, FieldInfo field, Type type, Type parentType)
        {
            var instance = (IBaseElement) field.GetValue(parent) ?? (IBaseElement) Activator.CreateInstance(type);
            var pageAttribute = field.GetAttribute<PageAttribute>();
            pageAttribute?.FillPage((WebPage) instance, parentType);
            return instance;
        }

        private IBaseElement GetInstanceElement(object parent, Type parentType, FieldInfo field,
            string driverName)
        {
            var instance = CreateChildFromFieldStatic(parent, parentType, field, driverName);
            instance.SetFunction(GetFunction(field));
            return instance;
        }

        private IBaseElement CreateChildFromFieldStatic(object parent, Type parentClass, FieldInfo field,
            string driverName)
        {
            var instance = (WebBaseElement) field.GetValue(parent);
            if (instance == null)
                instance = ActionWithException(
                    () => GetElementInstance(field, driverName), 
                    ex => $"Can't create child for parent '{parentClass.Name}' with type '{field.FieldType.Name}'. Exception: {ex}");
            else if (instance.Locator == null)
                instance.WebAvatar.ByLocator = GetNewLocator(field);
            instance.Frame = FrameAttribute.GetFrame(field);
            return instance;
        }

        private WebBaseElement GetElementInstance(FieldInfo field, string driverName)
        {
            var type = field.FieldType;
            var fieldName = field.Name;
            var newLocator = GetNewLocator(field);
            return ActionWithException(() => 
            {
                WebBaseElement instance = null;
                if (type == typeof(List))
                    throw Exception($"Can't init element {fieldName} with type 'List<>'. Please use 'IList<>' or 'Elements<>' instead");
                if (typeof(IList).IsAssignableFrom(type))
                {
                    var elementClass = type.GetGenericArguments()[0];
                    if (elementClass != null)
                        instance = (WebBaseElement) Activator.CreateInstance(typeof(Elements<>)
                            .MakeGenericType(elementClass));
                }
                else  {
                    if (type.IsInterface)
                        type = MapInterfaceToElement.ClassFromInterface(type);
                    if (type != null)
                    {
                        instance = (WebBaseElement) Activator.CreateInstance(type);
                        instance.WebAvatar.ByLocator = newLocator;
                    }
                }
                if (instance == null)
                    throw Exception("Unknown interface: " + type +
                                    ". Add relation interface -> class in VIElement.InterfaceTypeMap");
                instance.Avatar.DriverName = driverName;
                return instance;
            }, ex => $"Error in getElementInstance for field {fieldName}' with type '{type.Name + ex.FromNewLine()}'");
        }

        private static By GetNewLocator(FieldInfo field)
        {
            return ActionWithException(() => 
            {
                var locatorGroup = AppVersion;
                if (locatorGroup != null)
                {
                    string groupName;
                    By byLocator;
                    JFindByAttribute.Get(field, out byLocator, out groupName);
                    if (groupName != null && locatorGroup.Equals(groupName) && byLocator != null)
                        return byLocator;
                }
                return FindByAttribute.FindsByLocator(field) ?? FindByAttribute.Locator(field);
            }, ex => $"Error in get locator for type '{field.Name + ex.FromNewLine()}'");
        }

    }
}
